cpp/log/cert.cc

```diff
@@ -546,12 +566,314 @@ Cert::Status Cert::ExtensionStructure(int extension_nid,
+    if (str == "?") {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool Cert::IsValidRedactedHost(const std::string& hostname) {
+  // Split the hostname on '.' characters
+  std::vector<string> tokens = util::split(hostname, '.');
+
+  // Enforces the following rules: '?' must be to left of non redactions
+  // If first label is '*' then treat it as if it was a redaction
+  bool can_redact = true;
+  for (int pos = 0; pos < tokens.size(); ++pos) {
+    if (tokens[pos] == "?") {
+      if (!can_redact) {
+        return false;
+      }
+    } else {
+      // Allow a leading '*' for redaction but once we've seen anything else
+      // forbid further redactions
+      if (tokens[pos] != "*") {
+        can_redact = false;
+      } else if (pos > 0) {
+        // '*' is only valid at the left
+        return false;
+      }
+    }
+  }
+
+  return true;
+}
+
+Cert::Status Cert::IsValidWildcardRedaction() const {
+  if (!IsLoaded()) {
+    LOG(ERROR) << "Cert not loaded";
+    return Cert::ERROR;
+  }
+
+  std::string common_name;
+  std::vector<std::string> dns_alt_names;
+  int redacted_name_count = 0;
+
+  // First. Check all the Subject Alt Name extension records. Any that are of
+  // type DNS must pass validation if they are attempting to redact labels
+  STACK_OF(GENERAL_NAME) * subject_alt_names = (STACK_OF(GENERAL_NAME) *)
+      X509_get_ext_d2i(x509_, NID_subject_alt_name, NULL, NULL);
+
+  if (subject_alt_names != NULL) {
+    int subject_alt_name_count = sk_GENERAL_NAME_num(subject_alt_names);
+
+    for (int i = 0; i < subject_alt_name_count; ++i) {
+      GENERAL_NAME* name = sk_GENERAL_NAME_value(subject_alt_names, i);
+
+      if (name -> type == GEN_DNS) {
+        std::string dns_name = std::string(
+            reinterpret_cast<char*>(ASN1_STRING_data(name->d.dNSName)),
+            ASN1_STRING_length(name->d.dNSName));
+
+        // Make sure there isn't an embedded NUL character in the DNS ID
+        if (ASN1_STRING_length(name->d.dNSName) != dns_name.length()) {
+          LOG(ERROR) << "Embedded null in DNS-ID";
+          sk_GENERAL_NAME_free(subject_alt_names);
+          return Cert::ERROR;
+        }
+
+        dns_alt_names.push_back(dns_name);
+
+        if (IsRedactedHost(dns_name)) {
+          if (!IsValidRedactedHost(dns_name)) {
+            LOG(WARNING) << "Invalid redacted host: " << dns_name;
+            sk_GENERAL_NAME_free(subject_alt_names);
+            return Cert::FALSE;
+          }
+
+          redacted_name_count++;
+        }
+      }
+    }
+
+    sk_GENERAL_NAME_free(subject_alt_names);
+  }
+
+  // The next stage of validation is that if the subject name CN exists it
+  // must match the first DNS id and have the same labels redacted
+  // TODO(mhs): Confirm it's valid to not have a CN.
+  X509_NAME* name = X509_get_subject_name(x509_);
+
+  if (name == NULL) {
+    LOG(ERROR) << "Missing X509 subject name";
+    return Cert::ERROR;
+  }
+
+  int name_pos = -1;
```

Comment:
I think it does but I'm returning a different status for the two cases. Arguably they're the same.